packages/template-retail-react-app/app/utils/url.js

```diff
@@ -150,22 +151,83 @@ export const getPathWithLocale = (shortCode, opts = {}) => {
     )
     return newUrl
 }
 
+/**
+ * Generates the URL Template literal (Template string) used to build URLs in the App according
+ * the current selected site/locale and the default App URL configuration.
+ *
+ * @param appConfig Application default configuration.
+ * @param siteRef Current selected Site reference. The value can be the Site id or alias.
+ * @param localeRef Current selected Locale reference. The value can be the Locale id or alias.
+ * @returns {function(*, *, *, *=): string} function providing href, site and locale generates a site URL.
+ */
+export const createUrlTemplate = (appConfig, siteRef, localeRef) => {
+    const {site: siteConfig, locale: localeConfig, showDefaults: showDefaultsConfig} = appConfig.url
+    const defaultSite = getDefaultSite()
+    const sites = getSites()
+    const site =
+        sites.find((site) => {
+            return site.alias === siteRef || site.id === siteRef
+        }) || defaultSite
+    const defaultLocale = getLocaleByReference(site, site.l10n.defaultLocale)
+
+    const isDefaultSite =
+        defaultSite.id === siteRef || (defaultSite.alias && defaultSite.alias === siteRef)
+    const isDefaultLocale =
+        defaultLocale.id === localeRef || (defaultLocale.alias && defaultLocale.alias === localeRef)
+
+    const querySite =
+        (siteConfig === urlPartPositions.QUERY_PARAM && showDefaultsConfig) ||
+        (siteConfig === urlPartPositions.QUERY_PARAM && !showDefaultsConfig && !isDefaultSite)
+    const queryLocale =
+        (localeConfig === urlPartPositions.QUERY_PARAM && showDefaultsConfig) ||
+        (localeConfig === urlPartPositions.QUERY_PARAM && !showDefaultsConfig && !isDefaultLocale)
+
+    const isQuery = querySite || queryLocale
+    const isQuerySiteAndLocale = querySite && queryLocale
+
+    const pathSite =
+        (siteConfig === urlPartPositions.PATH && showDefaultsConfig) ||
+        (siteConfig === urlPartPositions.PATH && !showDefaultsConfig && !isDefaultSite)
+    const pathLocale =
+        (localeConfig === urlPartPositions.PATH && showDefaultsConfig) ||
+        (localeConfig === urlPartPositions.PATH && !showDefaultsConfig && !isDefaultLocale)
+
+    const isPathSiteAndLocale = pathSite && pathLocale
+
+    const templateConfig = {
+        pathSite,
+        pathLocale,
+        isPathSiteAndLocale,
+        querySite,
+        queryLocale,
+        isQuery,
+        isQuerySiteAndLocale
+    }
+
+    return (href, site, locale, config = templateConfig) =>
+        `${config.pathSite && site ? `/${site}` : ''}${
+            config.pathLocale && locale ? `/${locale}` : ''
+        }` +
+        `${href}` +
+        `${config.isQuery && (site || locale) ? '?' : ''}${
+            config.querySite && site ? `site=${site}` : ''
+        }${config.isQuerySiteAndLocale && locale ? '&' : ''}${
+            config.queryLocale && locale ? `locale=${locale}` : ''
+        }`
```

Comment:
@adamraya My attempt to make this code more readable.

```suggestion
    return (href, site, locale, config = templateConfig) => {
        const site = config.pathSite && site ? `/${site}` : ''
        const locale = config.pathLocale && locale ? `/${locale}` : ''
        const searchParams = new URLSearchParams()
        const hasQuery = config.isQuery && (site || locale)
        let queryString = ''

        if (hasQuery) {
            config.querySite && site && query.append('site', site)
            config.queryLocale && locale && query.append('locale', locale)
            queryString = `?${searchParams.toString()}`
        }

        return `${site}${locale}${href}${queryString}`
    }
```

Reply:
(I didn't go through the entire PR yet) 

In this code block, the variable naming is kinda weird here... why would you have href after site/locale?

Reply:
Thanks @kevinxh I tweaked a bit the suggestion. I agree makes the code a bit more readable. 
I also renamed `href` to `path`.